### ISSUE TITLE
For ConcatV2, `axis` really require int64 support?

### ISSUE THREAD

**Issue by shahidhs-ibm** [tensorflow/tensorflow#28114]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04, 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary):  source
- TensorFlow version (use command below): 0.12.0
- Python version: 2.7.15rc1
- Bazel version (if compiling from source): 0.15.0
- GCC/Compiler version (if compiling from source): 7.3.0
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the current behavior**
The `testConcatAxisType` test in `concat_op_test` is failing on an assertion for s390x:
```
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/87c06bad5aec2f21ec96b7253de551e8/execroot/org_tensorflow/bazel-out/s390x-opt/bin/tensorflow/python/kernel_tests/concat_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/concat_op_test.py", line 621, in testConcatAxisType
    self.assertEqual([2, 6], c.get_shape().as_list())
AssertionError: Lists differ: [2, 6] != [4, 3]
```

**Describe the expected behavior**
Assert values should match.

**Code to reproduce the issue**
On Intel:
```
>>> from __future__ import absolute_import
>>> from __future__ import division
>>> from __future__ import print_function
>>> import numpy as np
>>> import tensorflow as tf
>>> from tensorflow.python.framework import constant_op
>>> from tensorflow.python.framework import dtypes
>>> from tensorflow.python.ops import gen_array_ops
>>>
>>> t1 = [[1, 2, 3], [4, 5, 6]]
>>> t2 = [[7, 8, 9], [10, 11, 12]]
>>>
>>> c = constant_op.constant(1, dtype=dtypes.int32)
>>> z = gen_array_ops.concat_v2([t1, t2], c)
>>> z
<tf.Tensor 'ConcatV2_2:0' shape=(2, 6) dtype=int32>
>>>
>>> c = constant_op.constant(1, dtype=dtypes.int64)
>>> z = gen_array_ops.concat_v2([t1, t2], c)
>>> z
<tf.Tensor 'ConcatV2_3:0' shape=(2, 6) dtype=int32>
>>>
```

On s390x:
```
>>> from __future__ import absolute_import
>>> from __future__ import division
>>> from __future__ import print_function
>>> import numpy as np
>>> import tensorflow as tf
>>> from tensorflow.python.framework import constant_op
>>> from tensorflow.python.framework import dtypes
>>> from tensorflow.python.ops import gen_array_ops
>>>
>>> t1 = [[1, 2, 3], [4, 5, 6]]
>>> t2 = [[7, 8, 9], [10, 11, 12]]
>>>
>>> c = constant_op.constant(1, dtype=dtypes.int32)
>>> z = gen_array_ops.concat_v2([t1, t2], c)
>>> z
<tf.Tensor 'ConcatV2_2:0' shape=(2, 6) dtype=int32>
>>>
>>> c = constant_op.constant(1, dtype=dtypes.int64)
>>> z = gen_array_ops.concat_v2([t1, t2], c)
>>> z
<tf.Tensor 'ConcatV2_3:0' shape=(4, 3) dtype=int32>
>>>
```

**Other info / logs**
From above logs we can see that when we pass a constant (axis) of int64 type to `ConcatV2`, we get different results on s390x which causes assertion failure. 

After searching the historical records, I found this PR #14251 which has introduced this test case. However when I scanned the TensorFlow code for checking how the `concat` operations work, I found that it deals mostly with int32 type.

Getting different results on s390x for 32 Vs 64 bit data types is nothing new. We have handled couple of cases in issue #11431 and #14017 and it happens because of NumPy's different behavior on Intel Vs Z (PR #12963) . As my code understanding with `concat` is different, I am curious to understand if we really need to process int64 type of axis (is it a real time use case?) or its just added to fill the gap between `ConcatV2` supported types Vs its kernel implementation?
I might be wrong but want to be sure that I proceed to find fix for the right thing. Please check and confirm.

**Comment by saxenasaurabh** [comment]
@mrry @aselle do you have any ideas?

**Comment by mrry** [comment]
I think this was fixed by the change to concat_op.cc in https://github.com/tensorflow/tensorflow/commit/2d0531d72c7dcbb0e149cafdd3a16ee8c3ff357a, which added [these lines](https://github.com/tensorflow/tensorflow/blame/f4815551dc19b2149a70f20dba5a6c043e627a5b/tensorflow/core/kernels/concat_op.cc#L65-L87). That change isn't part of TensorFlow 0.12.0, but it looks like it became available with 1.8.0.

**Comment by shahidhs-ibm** [comment]
@mrry The concat_op_test is failing for the architecture I am working on and we are hitting the same numpy bug which we encountered in [#12963](https://github.com/tensorflow/tensorflow/pull/12963). In this issue I just wanted to check if we can skip int64 support for my architecture if its not that critical.

If I change the dtype to int32 in [make_tensor_proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/tensor_util.py#L376), the test passes for us. I need to add following code after [line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/tensor_util.py#L485):
```
  if nparray.dtype == np.int64:
    nparray = np.array(values, dtype=np.int32)
    dtype = dtypes.int32
```
Please suggest if you are aware of any better approach to achieve this. Thanks in advance :)
